rpcs3/Emu/RSX/rsx_methods.cpp

```diff
@@ -819,36 +819,85 @@ namespace rsx
 			{
-				if (index >= method_registers.nv308a_size_out_x())
+				const u32 out_x_max = method_registers.nv308a_size_out_x();
+
+				if (index >= out_x_max)
 				{
 					// Skip
 					return;
 				}
 
-				u32 color = arg;
-				u32 write_len = 4;
+				// Get position of the current command arg
+				const u32 src_offset = rsx->fifo_ctrl->get_pos() - 4;
+
+				// Get real args count (starting from NV3089_COLOR)
+				const u32 count = std::min<u32>({rsx->fifo_ctrl->get_remaining_args_count() + 1,
+					static_cast<u32>(((rsx->ctrl->put & ~3ull) - src_offset) / 4), 0x700 - index, out_x_max - index});
+
+				const u32 dst_dma = method_registers.blit_engine_output_location_nv3062();
+				const u32 dst_offset = method_registers.blit_engine_output_offset_nv3062();
+				const u32 out_pitch = method_registers.blit_engine_output_pitch_nv3062();
+
+				const u32 x = method_registers.nv308a_x() + index;
+				const u32 y = method_registers.nv308a_y();
+
+				// TODO
+				//auto res = vm::passive_lock(address, address + write_len);
+
 				switch (method_registers.blit_engine_nv3062_color_format())
 				{
 				case blit_engine::transfer_destination_format::a8r8g8b8:
 				case blit_engine::transfer_destination_format::y32:
 				{
-					// Bit cast
+					// Bit cast - optimize to mem copy
+
+					const auto dst = vm::_ptr<u8>(get_address(dst_offset + (x * 4) + (out_pitch * y), dst_dma, HERE));
+					const auto src = vm::_ptr<const u8>(get_address(src_offset, CELL_GCM_LOCATION_MAIN, HERE));
+
+					const u32 data_length = count * 4;
+					const bool is_overlapping = (dst_dma & CELL_GCM_LOCATION_MAIN) && [&]() -> bool
+					{
+						const u32 src_max = src_offset + data_length;
+						const u32 dst_max = dst_offset + (out_pitch * y) + data_length;
+						return (src_offset >= dst_offset && src_offset < dst_max) ||
+						 (dst_offset >= src_offset && dst_offset < src_max);
+					}();
+
+					if (is_overlapping)
+					{
+						for (u32 i = 0; i < count; i++)
+						{
+							reinterpret_cast<u32*>(dst)[i] = reinterpret_cast<const u32*>(src)[i];
+						}
+					}
+					else
+					{
+						std::memcpy(dst, src, data_length);
+					}
+
 					break;
 				}
 				case blit_engine::transfer_destination_format::r5g6b5:
 				{
-					// Input is considered to be ARGB8
-					u32 r = (arg >> 16) & 0xFF;
-					u32 g = (arg >> 8) & 0xFF;
-					u32 b = arg & 0xFF;
-
-					r = u32(r * 32 / 255.f);
-					g = u32(g * 64 / 255.f);
-					b = u32(b * 32 / 255.f);
-					color = (r << 11) | (g << 5) | b;
-					write_len = 2;
+					const auto dst = vm::_ptr<u16>(get_address(dst_offset + (x * 2) + (y * out_pitch), dst_dma, HERE));
+					const auto src = vm::_ptr<const u32>(get_address(src_offset, CELL_GCM_LOCATION_MAIN, HERE));
+
+					for (u32 i = 0; i < count; i++)
+					{
+						// Input is considered to be ARGB8
+						const u32 input = src[i];
+						u32 r = (input >> 16) & 0xFF;
+						u32 g = (input >> 8) & 0xFF;
+						u32 b = input & 0xFF;
+
+						r = ::rounded_div(r * 32, 255);
```

Comment:
Nope, not rounded but truncated. This might seem minor but is actually very destructive for shader parameters.

Reply:
hmm, it was rounded before because of float usage, was it wrong?

Reply:
rsx thread should not be rounding on int conversion. If it was/is rounding up, that is wrong.